tests: removing intg/ts_cache.py

danlavu opened the pull request:
the following test cases are now covered in system/test_cache.py and this can be removed.
 
* fixed assertion writes_to_both_databases tests
* added test detecting modification and deletion for groups
  * test is a common user story and functional, changed priority to critical
* added "integration" test invalidating user, group, netgroup objects

shridhargadekar commented:
> The following test cases are covered in system/test_cache.py

Could you share where exactly these tests getting covered in system/test_cache.py?
Some of old test cases have set up related to rfc2307bis, where these scenarios are being covered

shridhargadekar reviewed (changes requested):
Could you share where exactly these rfc2307bis related tests getting covered in system/test_cache.py?

alexey-tikhonov commented:
CI is red

danlavu commented:
@shridhargadekar It's not. We're not porting tests, the coverage is related to cache timestamp in ldb which is covered. 

I asked the team how to determine what LDAP tests we should test against rfc2307 and rfc2307bis, but we can easily take any test and parameterize the test to re-run it against 2307bis, so this can easily be done later.

shridhargadekar commented:
@danlavu not all tests have the coverage, [L655](https://github.com/SSSD/sssd/pull/7580/files#diff-ce47377d35b9c0d819ff86805a9262d349e4b9170c53aadcc0c6efe506e9c56bL655) [L628](https://github.com/SSSD/sssd/pull/7580/files#diff-ce47377d35b9c0d819ff86805a9262d349e4b9170c53aadcc0c6efe506e9c56bL628)
These scenarios do not have coverage.
The way these tests running is bit different from what [tests/test_cache.py](https://github.com/SSSD/sssd/blob/master/src/tests/system/tests/test_cache.py)

shridhargadekar commented:
If missing coverage is done later then consider approval from me.

danlavu commented:
I admit, I didn't take a super close look at the test cases. I'm going to go through them today and will probably end up writing tests for them.

danlavu commented:
I added two tests that cover the missing scenarios you mentioned.

Test that ensures the group changes are reflected after a clear
Another test, that ensures the cache records the change when that object type is cleared. 

Updated the assertion for one of the tests that was confusing, mentioned in slack.

danlavu commented:
Tests are not green, looking into it.

danlavu commented:
The test was not working correctly, it was passing when there were no results, fixed now.

alexey-tikhonov commented:
`test_colliding_hashes()` doesn't seem to be ported.
That's a question, though, if we want to keep it...

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,66 +147,159 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+        for k, v in result.items():
+            if isinstance(k, dict):
+                expire_time = result.get(str(k)).get("dataExpireTimestamp")
+                client.host.conn.exec(["sss_cache", f"-{obj[0]}", obj])
```
First 'sss_cache' (with 'dbs == cache') invalidates entry in both databases.
So when test gets to 'dbs == timestamps', entry should be already invalidated and 'expire_time' should be already reset.
Not sure how it works..?

danlavu replied:
Yes, it's supposed to. The expire_time variable contains the value before clearing the cache and it's compared against the new dataExpireTimestamp.

alexey-tikhonov replied:
On a 2nd run - when 'dbs == timestamps' - 'expire_timestamp' is set to a value from a cache that was already cleared on a previous run, when 'dbs == cache'.

alexey-tikhonov replied:
@danlavu, ping

danlavu replied:
Sorry, this was put on the back burner. Yes, I see the issue now.  I also added another assertion to ensure that the value is not empty, it was skipping the assertion and passing test because no assertion was made.

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,66 +147,166 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+@pytest.mark.importance("low")
+@pytest.mark.parametrize("obj", ["user", "group", "netgroup"])
+@pytest.mark.parametrize("dbs", ["cache", "timestamps"])
+@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
```
I suggest running this "low importance" test against LDAP only.

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,66 +147,166 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+@pytest.mark.integration
+@pytest.mark.importance("low")
+@pytest.mark.parametrize("obj", ["user", "group", "netgroup"])
+@pytest.mark.parametrize("dbs", ["cache", "timestamps"])
```
I understand this is more elegant way as compared to searching two DBs in a single test run.
But mind the CI overhead...

danlavu commented:
@alexey-tikhonov, I converted all the tests to use the LDAP provider, as we discussed, and I corrected the logic; it's working now and not being skipped. I did notice that netgroups in the timestamp cache is empty. There is nothing in the results. Do you know if netgroups is supposed to write to the timestamp cache?

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,70 +146,174 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+
+    path = f"/var/lib/sss/db/{dbs}_{client.sssd.default_domain}.ldb"
+
+    # netgroups does not use the fully qualified name and does not use the timestamp cache
```
From a quick glance, netgroups /should/ be using ts cache as well:
`sysdb_search_netgroup_by_name()` -> `sysdb_search_entry()` -> `sysdb_cache_search_entry()` + `sysdb_merge_msg_list_ts_attrs()`

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,70 +146,174 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+    assert result != {}, f"ldbsearch {ldb_filter} did not return any results!"
+    for _, v in result.items():
+        result_expire_time = v.get("dataExpireTimestamp")
+        if result_expire_time is not None:
```
Better assert it is not 'None'.

danlavu replied:
Hrmm.... 

```
(Pdb) client.tools.getent.netgroup("netgroup")
netgroup (-,user,))

(Pdb) client.ldb.search(path, suffix)
{}

(Pdb) client.ldb.search(f"/var/lib/sss/db/cache_{client.sssd.default_domain}.ldb", suffix)
{'name=netgroup,cn=Netgroups,cn=test,cn=sysdb': {'dn': ['name=netgroup,cn=Netgroups,cn=test,cn=sysdb'], 'createTimestamp': ['1729646403'], 'name': ['netgroup'], 'objectClass': ['netgroup'], 'originalDN': ['cn=netgroup,ou=netgroups,dc=ldap,dc=test'], 'originalModifyTimestamp': ['20241023012002Z'], 'netgroupTriple': ['(-, user, )'], 'lastUpdate': ['1729646403'], 'dataExpireTimestamp': ['1729651803'], 'distinguishedName': ['name=netgroup,cn=Netgroups,cn=test,cn=sysdb']}}

(Pdb) client.ldb.search(f"/var/lib/sss/db/timestamps_{client.sssd.default_domain}.ldb", suffix)
{}

I checked it against the client system, and the object doesn't exist in the timestamps db. 
```

alexey-tikhonov replied:
I guess you replied to a wrong comment.
But this might be a real bug.

alexey-tikhonov commented on src/tests/system/tests/test_cache.py:
```diff
@@ -146,70 +146,172 @@ def test_cache__writes_to_both_database_files_when_using_fully_qualified_names(
+        ldb_filter = f"dn=name={obj}@{client.sssd.default_domain},{suffix}"
+
+    result = client.ldb.search(path, suffix, filter=ldb_filter)
+    assert result != {}, f"ldbsearch {ldb_filter} did not return any results!"
```
Let's also print 'path' or 'dbs' here (and in a 2nd run).

alexey-tikhonov commented:
@shridhargadekar, please review the latest version.

danlavu commented:
The failure is not related to the patch.

alexey-tikhonov commented:
@danlavu, there is a conflict in sssd-2-9, please open explicit backport PR if needed.

alexey-tikhonov commented:
Pushed PR: https://github.com/SSSD/sssd/pull/7580

* `master`
    * d5b648498a712db4287c0c1050a0bc415ff49948 - tests: converting all the ldb cache tests to use one provider
    * be0c232be01b5fe2ffcaecae84980fb1b67ab916 - tests: removing intg/ts_cache.py
* `sssd-2-10`
    * 0ceefae810022e61f37b3a6dc6351018f46dd6db - tests: converting all the ldb cache tests to use one provider
    * c6b9e2645a75bc88283c1d8bff58c879777fc45f - tests: removing intg/ts_cache.py
